[CA-965] delete google project when deleting v2 workspaces

andy7i opened the pull request:
Tests to create a v2 workspace and delete the workspace (including the google project) were verified using the integration test running against a fiab with the changes in this PR and in #1330, as well as https://github.com/broadinstitute/firecloud-develop/pull/2354. I've reverted those changes so that they can be reviewed separately.

As part of this ticket, doug helped set up the Rawls SA to have google project delete IAM in the rbs project folder. [These SA permissions (and others)](https://broadinstitute.slack.com/archives/CNXF2BAPJ/p1608155536467900?thread_ts=1608072049.433700&cid=CNXF2BAPJ) will be saved in terraform by devops (probably?)


- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

yonghaoy committed:
upgarde RBS client library

andy7i committed:
Delete google project if v1 workspace

andy7i committed:
fix / add unit test

andy7i committed:
unit test for delete v1 workspace

andy7i committed:
sync the billing project owner policy group
hopefully fixes this issue:
I can create v2 workspaces in a v1 billing project but not in a v2 billing project and I'm not seeing much in the fiab logs. The error I'm seeing is:
```
[INFO] [16:18:20.570] [scala-execution-context-global-867] o.b.dsde.rawls.dataaccess.HttpSamDAO - retry-able operation failed: retries remain but predicate failed, not retrying
spray.json.JsonParser$ParsingException: Unexpected end-of-input at input index 0 (line 1, position 1), expected JSON Value:
```
slack thread with even more details:
 https://broadinstitute.slack.com/archives/CNXF2BAPJ/p1607531509347100

andy7i committed:
fix unit tests

andy7i committed:
fix a whoopsie (wrong resource type)

andy7i committed:
update unit tests

andy7i committed:
new integration test

andy7i committed:
comments

andy7i committed:
Merge branch 'v2_workspaces' into aj-delete-projects-in-v2-workspaces--ca-965

# Conflicts:
#	core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

andy7i committed:
update unit test to fix merge issue

andy7i commented on automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspaceApiSpec.scala:
```diff
@@ -62,6 +62,30 @@ class WorkspaceApiSpec extends TestKit(ActorSystem("MySpec")) with AnyFreeSpecLi
+        Rawls.workspaces.delete(billingProjectName, workspaceName)
+        assertNoAccessToWorkspace(billingProjectName, workspaceName)
+
+        Rawls.workspaces.delete(billingProjectName, workspaceCloneName)
+        assertNoAccessToWorkspace(billingProjectName, workspaceCloneName)
```
is it possible to verify that a google project has been deleted? If so, I think this would require:
- updating rawls / workspace model in workbench-libs  to get google-project-id (which it is missing and therefore errors) -- we probably need to do this eventually anyway to get all of the tests working
- making a call to google directly (unless we have some API that can do this)

dvoet replied:
there is a google api to check this and I agree we should update the model

andy7i replied:
updated the model and started a workbench-libs branch for v2_workspaces here: https://github.com/broadinstitute/workbench-libs/tree/v2_workspaces

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -437,6 +441,38 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+
+  private def deletePetsInProject(projectName: GoogleProjectId, userInfo: UserInfo): Future[Unit] = {
+    for {
+      projectUsers <- samDAO.listAllResourceMemberIds(SamResourceTypeNames.googleProject, projectName.value, userInfo)
```
when I try to delete the v2 workspaces the call errors: 

```
{"causes":[],"message":"http error calling uri https://sam-fiab.dsde-dev.broadinstitute.org/api/resources/v1/google-project/terra-test-0e0f9a61/allUsers","source":"rawls","stackTrace":[],"statusCode":404}
```

BUT the workspace does get deleted

do we need to create an allUsers group for the billing project resource in sam? Is this another scenario where we assume a group has been synced but hasn't been synced yet?

dvoet replied:
this is hitting this api
https://sam.dsde-prod.broadinstitute.org/#/Resources/getAllResourceUsers
there are no groups involved

andy7i replied:
Thanks Doug! I looked further into the 404 and it looks like when we create v2 workspaces, we never create the sam resources for the google-project. I'm working on fixing this now (and may stick this fix into a new PR)

andy7i replied:
PR to fix this issue: https://github.com/broadinstitute/rawls/pull/1330

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -437,6 +441,38 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
 
+  // TODO - once workspace migration is complete and there are no more v1 workspaces or v1 billing projects, we can remove this https://broadworkbench.atlassian.net/browse/CA-1118
+  private def maybeDeleteGoogleProject(googleProjectId: GoogleProjectId, workspaceVersion: WorkspaceVersion, userInfoForSam: UserInfo): Future[Unit] = {
+    val isV1Workspace: Boolean = workspaceVersion == WorkspaceVersions.V1
+    if (!isV1Workspace)
```
I would probably have
```
if (workspaceVersion == WorkspaceVersions.V2)
```
and please use curly braces

andy7i committed:
PR feedback

andy7i committed:
add missing google-project sam resource.
while working on CA-965, I found that I was getting 404s from the Sam allUsers endpoint.

andy7i committed:
disambiguation: rename `SamProjectRoles` to `SamBillingProjectRoles`

andy7i committed:
Delete the google project using the Rawls SA instead of the billing SA, in order to be more consistent about which SA owns what (billing SA owns billing; rawls SA owns user data and resources. https://broadinstitute.slack.com/archives/CNXF2BAPJ/p1608141734441200?thread_ts=1608072049.433700&cid=CNXF2BAPJ

andy7i committed:
handle delete google project based on whether it's a v1 billing project deletion or a v2 workspace deletion
V2 workspace projects are managed by rawls SA but the v1 billing google projects are managed by the billing SA.

andy7i committed:
update workbench-libs

andy7i committed:
~draw the rest of the owl~ do the rest of the test

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -945,11 +945,38 @@ class HttpGoogleServicesDAO(
   }
 
-  override def deleteProject(googleProject: GoogleProjectId): Future[Unit]= {
+  // TODO - once workspace migration is complete and there are no more v1 workspaces or v1 billing projects, we can remove this https://broadworkbench.atlassian.net/browse/CA-1118
+  // V2 workspace projects are managed by rawls SA but the v1 billing google projects are managed by the billing SA.
+  override def deleteV1Project(googleProject: GoogleProjectId): Future[Unit]= {
```
looks like github struggled a bit with the diff here.

The changes listed here were implemented so that it would be simpler and cleaner to rip out v1 billing google project deletion once migration is complete:
- deleteProject was renamed to deleteV1Project. this is used by v1 billing, to delete the google project. Those google projects cannot be managed by the rawls SA.
- a copy of deleteProject was made (to be used moving forward, with v2 workspaces, which uses the rawls SA creds.

andy7i committed:
Revert "disambiguation: rename `SamProjectRoles` to `SamBillingProjectRoles`"

This reverts commit 15403bd8

andy7i committed:
Revert "add missing google-project sam resource. while working on CA-965, I found that I was getting 404s from the Sam allUsers endpoint."

This reverts commit 06723063

s-rubenstein commented on automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspaceApiSpec.scala:
```diff
@@ -62,6 +64,46 @@ class WorkspaceApiSpec extends TestKit(ActorSystem("MySpec")) with AnyFreeSpecLi
+
+        implicit val ec: ExecutionContext = ExecutionContext.global
+        val source = scala.io.Source.fromFile(RawlsConfig.pathToQAJson)
+        val jsonCreds = try source.mkString finally source.close() // read in the json file and safely close it
```
I think the code here successfully says what it means, I don't think you need the comment.

s-rubenstein commented on automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspaceApiSpec.scala:
```diff
@@ -62,6 +64,46 @@ class WorkspaceApiSpec extends TestKit(ActorSystem("MySpec")) with AnyFreeSpecLi
         }
       }
 
+      "to create, clone, and delete v2 workspaces (using the Resource Buffer), in a v2 billing project" in {
```
Opt: This feels like a very chunky test. Would it make sense to break out `create` `delete` and `clone` to their own tests?

andy7i replied:
the main purpose of this test is to verify that deletion works. I'll rename the test to make that clear and I'll also remove the cloning to make it a bit smaller!

s-rubenstein commented on automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspaceApiSpec.scala:
```diff
@@ -62,6 +64,46 @@ class WorkspaceApiSpec extends TestKit(ActorSystem("MySpec")) with AnyFreeSpecLi
         }
       }
 
+      "to create, clone, and delete v2 workspaces (using the Resource Buffer), in a v2 billing project" in {
```
Is this a unit test or an integration test? Given that its in the automation folder, I assume integration?

andy7i replied:
yep!

s-rubenstein commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -945,11 +945,38 @@ class HttpGoogleServicesDAO(
+        executeGoogleRequest(resMgr.projects().delete(googleProject.value))
+      }) {
+        case e: GoogleJsonResponseException if e.getDetails.getCode == 403 && "Cannot delete an inactive project.".equals(e.getDetails.getMessage) => new Empty()
+        // stop trying to delete an already deleted project
```
I would expect a 403 from Google to mean insufficient access to the Google Project to delete. Given it should be using RBS projects, that speaks of a bug to me, not an already deleted project. I would expect a deleted project to give us a 404?

(Unless Google did a bad)

s-rubenstein replied:
(Although maybe this is existing code? That you just copied?)

andy7i replied:
yep it's copied over from existing code, but also I verified the behavior where you delete a project that was already "deleted" and it's actually just marked inactive.

s-rubenstein commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -945,11 +945,38 @@ class HttpGoogleServicesDAO(
+    }
+  }
+
+  override def deleteProject(googleProject: GoogleProjectId): Future[Unit]= {
```
Nit: Rename to `deleteGoogleProject`

andy7i replied:
I thought about renaming it, but thought that maybe within the context of the GoogleServicesDAO it might be ok. I should have just renamed it! 🙃

s-rubenstein commented on core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala:
```diff
@@ -598,6 +613,91 @@ class WorkspaceApiServiceSpec extends ApiServiceSpec {
+        }
+      }
+
+    verify(services.samDAO).deleteResource(
```
I would also verify that the gcsDao deleteProject is called. Just for completeness, and to avoid potentially leaving ghost projects behind.

s-rubenstein reviewed (approved):
Looks good! I dropped a few comments in there for you, but its tests and naming, so I am not going to block on them

andy7i committed:
pr feedback; making the test more clear and streamlined

andy7i committed:
pr feedback: verify that gcsDAO.deleteProject was called. Needed to add a spy on the mockGcsDao since mockito sees it as a real class.

andy7i committed:
pr feedback: rename deleteProject to deleteGoogleProject

andy7i commented:
verified integration test passed
